# **例題 1-1:** モンテカルロ法を用いた円周率$\pi$の計算
次のコードは，モンテカルロ法を用いて円の面積から円周率$\pi$を求めるコードである．

In [ ]:
import numpy as np
import math

sample = 10000
count = 0

for c in range(sample):
  x = np.random.rand()
  y = np.random.rand()
  if (x*x + y*y) <= 1.0: count += 1 

print(count/sample*4)

3.12


# **課題 1-1:** モンテカルロ法を用いたn次元球による円周率$\pi$の計算
例題1-1 を参考にして，n次元球の体積から円周率$\pi$を求める関数をプログラムしなさい．
なお$\Gamma$関数の計算には，math.gamma()を用いて良い．

In [ ]:
import numpy as np
import math

def calc_pi(n, sample):
  pos = 0 # 打たれた点の位置
  count = 0 # 球の中に打たれた点

  # sample回繰り返す
  for i in range(sample):
    # n次元分の計算
    for j in range(n):
      pos += np.random.rand() ** n
    # 球の中に打たれたか否か
    if (pos <= 1.0):
      count += 1
    # pos リセット
    pos = 0
  
  # 2021_CGのための幾何学_01.pdf -> スライド19 公式参照
  ans = (((count / sample) * (2**n) * (math.gamma(n / 2 + 1))) ** (2 / n))

  return ans

課題1-1の確認用コード (適当に数値を変えながらデバックや実験に使うと良い)

In [ ]:
pi = calc_pi(10, 100000)
sample = 100000

for i in range(10):
  i += 1
  print(f"{i}次元の円周率: {calc_pi(i, sample)}")

1次元の円周率: 3.1415926535897936
2次元の円周率: 3.14856
3次元の円周率: 3.856220429585467
4次元の円周率: 4.646512670810228
5次元の円周率: 5.456059139038563
6次元の円周率: 6.255883270868786
7次元の円周率: 7.051495525656328
8次元の円周率: 7.859269929073194
9次元の円周率: 8.637792252462491
10次元の円周率: 9.43206807892877


# **課題 1-2:** 課題1-1の結果の考察
次元$n$を大きくすると，求まる円周率$\pi$の誤差は大きくなるように見える．その理由が何故なのか考察し，答えなさい．グラフを用いたり，追加のコードを書いても良い．

In [ ]:
# n次元 立方体, 球 の体積を求める。
import numpy as np
import math

def calc_sphereV(r, n):
  ans = (math.pi ** (n / 2) ) / math.gamma((n / 2) + 1) * r ** n
  return ans

def calc_cubeV(r, n):
  ans = (2 * r)**n
  return ans

r = 1
for n in range(15):
  n += 1
  print(f"{n}次元球の体積: {calc_sphereV(r, n)},  {n}次元立方体の体積: {calc_cubeV(r, n)},  {n}次元立方体に含まれる{n}次元球体の割合: {calc_sphereV(r, n) / calc_cubeV(r, n)}")

1次元球の体積: 1.9999999999999998,  1次元立方体の体積: 2,  1次元立方体に含まれる1次元球体の割合: 0.9999999999999999
2次元球の体積: 3.141592653589793,  2次元立方体の体積: 4,  2次元立方体に含まれる2次元球体の割合: 0.7853981633974483
3次元球の体積: 4.1887902047863905,  3次元立方体の体積: 8,  3次元立方体に含まれる3次元球体の割合: 0.5235987755982988
4次元球の体積: 4.934802200544679,  4次元立方体の体積: 16,  4次元立方体に含まれる4次元球体の割合: 0.30842513753404244
5次元球の体積: 5.263789013914325,  5次元立方体の体積: 32,  5次元立方体に含まれる5次元球体の割合: 0.16449340668482265
6次元球の体積: 5.167712780049969,  6次元立方体の体積: 64,  6次元立方体に含まれる6次元球体の割合: 0.08074551218828077
7次元球の体積: 4.7247659703314016,  7次元立方体の体積: 128,  7次元立方体に含まれる7次元球体の割合: 0.036912234143214075
8次元球の体積: 4.058712126416768,  8次元立方体の体積: 256,  8次元立方体に含まれる8次元球体の割合: 0.0158543442438155
9次元球の体積: 3.2985089027387064,  9次元立方体の体積: 512,  9次元立方体に含まれる9次元球体の割合: 0.006442400200661536
10次元球の体積: 2.550164039877345,  10次元立方体の体積: 1024,  10次元立方体に含まれる10次元球体の割合: 0.00249039457019272
11次元球の体積: 1.8841038793898994,  11次元立方体の体積: 2048,  11次元立方体に含まれる11次元球体の割合: 0.0009199725973583493
12次元球の体積: 1.3352627688545893,  12次元立方

# 実行結果
---
表1. 課題1-1を実行した時に得られた結果

|  次元    |  円周率                |  
| :------: | :--------------------  |  
|  1       |  3.1415926535897936    |  
|  2       |  3.13252               |  
|  3       |  3.8616341135442993    |  
|  4       |  4.6475800154489       |  
|  5       |  5.4527865682948935    |  
|  6       |  6.257812353339532     |  
|  7       |  7.060951419113433     |  
|  8       |  7.854551232916127     |  
|  9       |  8.651212718238076     |  
|  10      |  9.436194325255673     |  

<br>

# 考察
---
表1から、次元 n を大きくすると求まる円周率 π の誤差は大きくなっていることがわかる。
ではなぜこのようになるのか、という点について考察していく。  
今回πを求めるために用いたモンテカルロ法は、n次元立法体の中に点を打ちn次元球の中に入った点の数の比を求め、n次元立法体の体積を掛けるというものである。
ここで私が考えたことは、高次元になるにつれて同じ比率で体積が増え、立方体内の球の割合が同じであれば誤差はあまり出ないのではないか、ということである。そこで、まず球の体積と立方体の体積を求め、次に立方体の中の球が同じ割合で存在しているかを求める。  
<br>
表2. 1～15次元の球と立方体の体積, 立方体に含まれる球体の割合

|  次元    |  球の体積              |  立方体の体積     | 割合
| :------: | :--------------------  | :---------------- | :----------
|  1       |  1.9999999999999998    |  2                | 0.9999999999999999
|  2       |  3.141592653589793     |  4                | 0.7853981633974483
|  3       |  4.1887902047863905    |  8                | 0.5235987755982988
|  4       |  4.934802200544679     |  16               | 0.30842513753404244
|  5       |  5.263789013914325     |  32               | 0.16449340668482265
|  6       |  5.167712780049969     |  64               | 0.08074551218828077
|  7       |  4.7247659703314016    |  128              | 0.036912234143214075
|  8       |  4.058712126416768     |  256              | 0.0158543442438155
|  9       |  3.2985089027387064    |  512              | 0.006442400200661536
|  10      |  2.550164039877345     |  1024             | 0.00249039457019272
|  11      |  1.8841038793898994    |  2048             | 0.0009199725973583493
|  12      |  1.3352627688545893    |  4096             | 0.00032599188692738996
|  13      |  0.910628754783283     |  8192             | 0.00011116073666788122
|  14      |  0.5992645293207919    |  16384            | 3.657620418217724e-05
|  15      |  0.38144328082330436   |  32768            | 1.1640725122781505e-05

<br>

求めた結果が表2である。この結果を見ると、立方体の体積は増え続けているが球の体積は5次元を境に減っている。さらに、立方体に含まれる球体の割合を見てみると高次元になればなるほど減っていることがわかる。
このように割合が減っているということは、ランダムで点を打つときn次元球の中に点が打たれる確率が減っていると言える。つまり、割合が0に近づけば近づくほど誤差が大きくなると考えることができる。

<br>

よって、次元を大きくすると求まるπの誤差が大きくなる理由は以上であると考察した。

<br>

## 参考文献
---

*   ⼩池崇⽂ 「2021 CGのための幾何学 01 pdf」 https://cms.cis.k.hosei.ac.jp/pluginfile.php/31796/mod_resource/content/0/2021_CG%E3%81%AE%E3%81%9F%E3%82%81%E3%81%AE%E5%B9%BE%E4%BD%95%E5%AD%A6_01.pdf (2021/04/12 閲覧)


